tests/python_functional/src/driver_io/file/file_io.py

```diff
@@ -26,20 +26,16 @@
 class FileIO(File):
     def __init__(self, file_path):
         super(FileIO, self).__init__(file_path)
-        self.__readable_file = None
+        self.__readable_file = File(file_path)
         self.__writeable_file = File(file_path)
 
-    def read(self):
-        if not self.__readable_file:
-            self.__readable_file = self.open(mode="r")
+    def read(self, counter):
```

Comment:
Do we need a mandatory counter for read operation?
I see one use case where a mandatory `counter` in `FileIO::read` cannot be used: when a test with a file() destination wants to assert on the number of lines in a file (or wants to read the entire file).

In `functional_tests/logpath/test_multiple_flags.py` a pattern can be seen that the test reads exactly the same amount of lines from the beginning of the file that was previously written into it. But how about a misconfigured filter() in syslog-ng's config which leads that the destination file has more lines in it?
I would like to do the same which is used in SNMP related test cases, see below.

In `SNMPtrapd` (helpers/snmptrapd/conftest.py), `get_log` does a raw readline() without any wait_for_lines methods. The test asserts on if the expected log is in the read data.
See `functional_tests/destination_drivers/snmp_destination/v2c/test_snmp_destination_v2c_custom_community.py`:
```
    assert expected_traps == snmptrapd.get_traps(len(expected_traps))
    assert test_community in snmptrapd.get_log()
```
There's also a `get_traps` method which is similar to `FileIO:read` in a sens that the user must provide the expected number of results that is expected.
Of course `SNMPtrapd` uses `File` class, therefore it has flexibility.
(I'm not suggesting that SNMPtrapd should be using FileIO instead)

My first question can be answered if the purpose of `FileIO` class is answered.
I think `FileIO` provides a higher-level interface to its users, who needs more than `File` class's raw I/O operations (however File just got some wait_for_lines methods).
If that is the case `FileIO:read()` either should be renamed to `wait_for_lines` or it should do raw I/O operations as its name suggests.

Reply:
Thanks for this review comment, interesting question.
I am agreed with you in that for now we can not achieve such use-case for file destinations where we want to assert on that how many lines are written by syslog-ng (with the help of FileDestination class).
If I understand you correctly you did not pointing that this is a loosed feature from the past (this case was not covered also before this PR). Just for the note before this PR there was a function DestinationReader::read_all_logs which was used only by SNMPtrapd class.
Mainly this PR wanted to keep the original features for the related syslog-ng drivers: FileDestination and SNMPtrapd destination. If this is not applied please notice it.

The purpose of FileIO class is the following:
- it provides only IO related functions for file based syslog-ng drivers such: reading/writing files
- this class should not be used from non FileBased drivers such: SyslogNg ConsoleLogReader, or SNMPtrapd and so on
- this class should be used only in such syslog-ng destination or source drivers where syslog-ng reads from or writes to files
- from my point of view this class can have several read/write methods, what file based drivers can dictate, and in this case I also can imagine such a method: read_all() which would returns with the entire file content.


The purpose of File class is the following:
- this class provides such general functions which are related to native File objects not specially to syslog-ng drivers, where syslog-ng drivers are not affected
- can be used in a following places where syslog-ng drivers are not in context: ConsoleLogReader (reading sterr of syslog-ng), or SNMPtrapd (reading output of snmptrapd tool), SyslogNgConfig (writing syslog-ng config file)
- I think it is easier to think about this class such a general "helper" class for native file based objects

All in all I would extend later FileIO with read_all() method (as you suggest). For now without this read_all() SyslogNgCtl class can help to ensure if syslog-ng writes the correct number of lines, example:
```
assert snmp_destination.get_stats() == {'written': expected_message_counter, 'processed': expected_message_counter, 'dropped': 0, 'queued': 0}
```

Do you agreed with this concept, or was it helpfull?

Reply:
Thanks for discussing this with me thoroughly!

As you ask, one functionality did change with this PR: `FileIO:read()` methods definition has a mandatory `counter` argument. I'm not against this as you've stated the reasons behind it. 
```diff
-    def read(self):
-        if not self.__readable_file:
-            self.__readable_file = self.open(mode="r")
+    def read(self, counter):
```
As you mentioned, `FileIO` class can be extended with a `read_all` method when/if needed. It would simply call the underlying `File:read()` method.
Thanks, this is resolved by me.

Reply:
You are right, this part of FileIO API has changed.
But from the testcase side it "remains" same logic:

This is how before this PR file destination read has working:
```
file_destination:read_log() -> file_destination:read_logs(counter) -> destination_reader:read_logs(counter) -> message_reader:pop_messages(counter)
```

After this PR file destination read will work in this way:
```
file_destination:read_log() -> file_destination:read_logs(counter) -> fileio:read(counter)
```

Reply:
And about the `destination_reader:read_all_logs`, yes it has been removed, but in one of the next PR-s it will be added (as I count)